### DISCUSSION TITLE
cordova requirements Command failed with ENOENT

### DISCUSSION THREAD

**Discussion by AlainReve** [apache/cordova#469] [Q&A]
I'm really stuck with : 
Requirements check results for android:
Java JDK: installed 17.0.6
Android SDK: installed true
Android target: not installed 
Command failed with ENOENT: avdmanager list target
spawn avdmanager ENOENT                                                                                         
Gradle: installed /opt/gradle/gradle-7.6/bin/gradle
Some of requirements check failed

I found the list here : 
https://cordova.apache.org/docs/en/latest/guide/platforms/android/index.html

What I do does not seem to be recognised by Cordova. For example, 

"java -version" : openjdk version "19.0.2" 2023-01-17
(and I tried it up to version 21 - I've got 6 java versions on my system now!)
but cordova still only sees version 17. 

I installed sdkmanager, following instructions here : 
https://www.androidcentral.com/installing-android-sdk-windows-mac-and-linux-tutorial
then I typed: 
~/AndroidSDK/cmdline-tools/latest/bin$ ./sdkmanager --install "platforms;android-33"
(I installed from android-25 to android-33)
but cordova still says "Android target: not installed". I thought these "platforms;android-xx" WERE the targets.

I fiddled with $PATH and so-and-so's HOME environment variables. 

sdkmanager --list : lists around 400 packages and I don't know which to choose. I don't know what half of the things on https://cordova.apache.org/docs/en/latest/guide/platforms/android/index.html ARE, I try to look them up and I install random packages (via sdkmanager, apt and snap). I already managed to mess up my system once. Got it fixed (I hate it when apt upgrade returns errors). 

Hum... Any plans to create a cordova package sometimes soon? Then I could just type "sudo apt install cordova", eventually accept one or two licences along the way, and I would get a functional version with all required dependencies... As it stands, I feel I'm going round in circles... 

Why doesn't cordova notice when I change the java sdk version? It noticed when I upgraded Gradle. OK, "why" is not the right question, is it? How can I make it notice? How can I tell it there are android targets on my system (supposing the "platforms;android-xx" really are android targets"?

**Comment by breautek** [comment]
> Why doesn't cordova notice when I change the java sdk version?

Cordova first checks and uses `CORDOVA_JAVA_HOME` environment variable if set, then falls back to `JAVA_HOME` (if I recall correctly.)

So if `CORDOVA_JAVA_HOME` or `JAVA_HOME` is set to a JDK 17 environment, it will use those binaries over say a `PATH` order. I'd also note that JDK 17 would be the recommended environment to use as that's what Android Studio will use and their tools assumes a JDK 17 environment.  JDK 19 or later is not officially supported by the android SDK so mileage may vary there.

For what it's worth, I also get

```
Android target: not installed
Command failed with ENOENT: avdmanager list target
spawn avdmanager ENOENT
```

when running the requirements command, but cordova and android building still works, so those might be a bug with the requirements command itself, rather than the bug in the actual cordova-android platform.

> How can I tell it there are android targets on my system (supposing the "platforms;android-xx" really are android targets"?

A workaround would be to use `sdkmanager` directly as you were already doing:

`$ANDROID_HOME/cmdline-tools/latest/bin/sdkmanager --list_installed`

And you're right, you're looking for `platforms;...` lines:

```
  platforms;android-33                                  | 3            | Android SDK Platform 33                    | platforms/android-33                                 
  platforms;android-34                                  | 3            | Android SDK Platform 34                    | platforms/android-34  
```

In my case, I have both android API 33 and 34 installed. At this time, Cordova only supports API 33. API 34 support will be introduced in the upcoming cordova-android@13 release.

**Reply by partharoy-digicert** [reply]
I was facing the same issue and added a comment in a different thread for the fix,

[Fix: cordova requirements ENOENT issue ](https://github.com/apache/cordova-android/issues/926#issuecomment-2982582632)
